{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c3850be-a729-4c33-ba10-c796d209e4dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import gffutils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3ca43d5-7ed1-4be3-afd5-722afb0630a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genome\n",
    "for record in SeqIO.parse(snakemake.input[0], \"fasta\"):\n",
    "    print(record.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b24fa229-9290-4a71-8674-9239a720ed7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = SeqIO.to_dict(SeqIO.parse(snakemake.input[0], \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51ccbf9a-bece-40a4-948b-1a8660764bee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotation\n",
    "\n",
    "db = gffutils.FeatureDB(snakemake.input[1])\n",
    "\n",
    "db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84e54089-109b-419c-8741-608e31153b3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "list(db.featuretypes())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49f07583-6af1-40a4-8d5c-d1efbda6fb67",
   "metadata": {},
   "outputs": [],
   "source": [
    "list(db.features_of_type(\"gene\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "659c9eca-46bd-4a3c-8a5b-a6ece7f38fcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in db.execute(\"SELECT name FROM sqlite_master WHERE type='table';\"):\n",
    "    print(row['name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f63c6672-2a23-4c85-9e83-5f58744a13d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in db.execute(\"pragma table_info(features);\"):\n",
    "    print(row['name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba9861d1-d4b6-4d7c-ab0b-37955f76b3e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in db.execute(\"select id from features where id like '%AT2G19110%';\"):\n",
    "    print(row['id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17afaa2b-fbdc-4a34-be67-c7b244a5c3b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "list(db['gene-AT2G19110'].attributes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c021e1d-7a5c-4d65-8747-05ea05cf4d7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "hma4_annotation = db['gene-AT2G19110']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c294d0a6-58f0-4221-bf4b-e799dc4de4fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "hma4_annotation.seqid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d9ef868-a8d2-4e81-818e-6618f3e6239c",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome[hma4_annotation.seqid][hma4_annotation.start-1:hma4_annotation.end].seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0264e3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# start coding here\n",
    "\n",
    "annotation = gffutils.FeatureDB(snakemake.input[1])\n",
    "\n",
    "genome = SeqIO.to_dict(SeqIO.parse(snakemake.input[0], \"fasta\"))\n",
    "\n",
    "gene_annotation_id = f\"gene-{snakemake.wildcards[0].split('.')[0]}\"\n",
    "\n",
    "gene_annotation = annotation[gene_annotation_id]\n",
    "\n",
    "gene_sequence = genome[gene_annotation.seqid][gene_annotation.start-1:gene_annotation.end]\n",
    "\n",
    "df = pd.read_csv(snakemake.input[2])\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "314d82b6-469c-49bb-89c5-7c0f6815d8fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_substitution_dna_window(variant, gene_sequence, gene_annotation):\n",
    "    in_gene_position = variant[\"position\"] - gene_annotation.start\n",
    "\n",
    "    ref, alt = parse_dna_substitution(variant[\"amino acid change\"])\n",
    "\n",
    "    assert ref == gene_sequence[in_gene_position].upper()\n",
    "    \n",
    "    return str(gene_sequence[in_gene_position - 256 : in_gene_position + 256].seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77d73ede-07a6-4534-aff0-58f05d4b056c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils import seq1\n",
    "import re\n",
    "\n",
    "\n",
    "def parse_aa_substitution(HGVS_string):\n",
    "    prot_match = re.match(r\"p\\.([A-Z]{1}[a-z]{2})\\d+([A-Z]{1}[a-z]{2})\", HGVS_string)\n",
    "    reference, alternative = prot_match.groups()\n",
    "    return seq1(reference), seq1(alternative)\n",
    "\n",
    "\n",
    "def parse_dna_substitution(HGVS_string):\n",
    "    dna_match = re.search(r\"c\\..*([ATCG])>([ATCG])$\", HGVS_string)\n",
    "    reference, alternative = dna_match.groups()\n",
    "    return reference, alternative\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2580de4e-db10-4e3b-a7be-0e10a61a5c62",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_sequence_window(snp_effect_path, reference_fasta_path, annotation_db_path):\n",
    "    annotation = gffutils.FeatureDB(annotation_db_path)\n",
    "    genome = SeqIO.to_dict(SeqIO.parse(reference_fasta_path, \"fasta\"))\n",
    "    \n",
    "    gene_annotation_id = f\"gene-{snakemake.wildcards[0].split('.')[0]}\"\n",
    "    gene_annotation = annotation[gene_annotation_id]\n",
    "    \n",
    "    gene_sequence = genome[gene_annotation.seqid][gene_annotation.start-1:gene_annotation.end]\n",
    "    \n",
    "    df = pd.read_csv(snp_effect_path)\n",
    "\n",
    "    df = df.dropna(subset=[\"amino acid change\"])\n",
    "    df = df[ ~ df[\"amino acid change\"].str.contains(\"ins\", na=False) ]\n",
    "    df = df[ ~ df[\"amino acid change\"].str.contains(\"del\", na=False) ]\n",
    "\n",
    "\n",
    "    df[\"sequence_window\"] = df.apply(\n",
    "        lambda x: get_substitution_dna_window(x, gene_sequence, gene_annotation), axis=1\n",
    "    )\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "478271ca-b2ea-4953-9604-db81542f6715",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = add_sequence_window(snakemake.input[2], snakemake.input[0], snakemake.input[1])\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b06e25a-45eb-4a1f-a705-8529eac39a1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df[\"sequence_window\"] == \"-\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71db427a-445e-4e52-b99a-4a31dc67588b",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8185f845-e0ac-44ab-bf58-5559e03f5ebd",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos = 8279002 - gene_annotation.start\n",
    "pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e47bb086-c4f6-4f86-8167-48f4a85763e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_sequence[pos]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "806fb145-6245-41bb-91ad-bb7b72ab93a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "\n",
    "def add_surprise_bonus(df):\n",
    "    # Entropy H in base-4\n",
    "    df[\"H\"] = -(\n",
    "        df[[\"p_a\", \"p_c\", \"p_g\", \"p_t\"]]\n",
    "        .pipe(lambda p: p * (np.log2(p) / 2))\n",
    "        .sum(axis=1)\n",
    "    )\n",
    "\n",
    "    # Surprisal I in base-4\n",
    "    df[\"I\"] = -(np.log2(df[\"p_obs\"]) / 2)\n",
    "    df[\"surprise bonus\"] = df[\"H\"] - df[\"I\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "170d379e-f112-47a5-9a5a-3d27420a49e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = pd.DataFrame({\n",
    "    \"p_a\": [0.3],\n",
    "    \"p_c\": [0.3],\n",
    "    \"p_g\": [0.3],\n",
    "    \"p_t\": [0.1],\n",
    "    \"p_obs\": [0.1], \n",
    "})\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61adc227-1bbf-4ff3-bc9e-cb01c5e6a756",
   "metadata": {},
   "outputs": [],
   "source": [
    "add_surprise_bonus(test)\n",
    "\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d96a43b-547e-441d-9d87-c6bc9ea94d4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = pd.DataFrame({\n",
    "    \"p_a\": [0.4],\n",
    "    \"p_c\": [0.4],\n",
    "    \"p_g\": [0.1],\n",
    "    \"p_t\": [0.1],\n",
    "    \"p_obs\": [0.1], \n",
    "    \"p_ref\": [0.1], \n",
    "})\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce788bb8-b385-4d67-9ccb-36bc77252299",
   "metadata": {},
   "outputs": [],
   "source": [
    "add_surprise_bonus(test)\n",
    "\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "390bbb57-d656-4cd1-97e3-ffffde5435d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "gpn = np.log2(test[[\"p_a\", \"p_c\", \"p_g\", \"p_t\"]] / test[\"p_ref\"].to_numpy()[:, None])\n",
    "\n",
    "gpn\n",
    "        "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
